src/vcml/models/dma/pl330.cpp

```diff
@@ -0,0 +1,1144 @@
+
+namespace vcml::dma {
+using namespace mwr;
+
+enum dsr_bits {
+    DNS = bit(9), // Manager security status
+};
+typedef field<0, 4, u32> DSR_DMA_STATUS;
+typedef field<4, 5, u32> DSR_WAKEUP_EVENT;
+
+enum fsrd_bits {
+    FS_MGR = bit(0), // Manager fault state
+};
+
+typedef field<0, 8, u32> FSRC_FAULT_STATUS;
+
+enum ftrd_bits {
+    FTRD_MASK = pl330::manager::UNDEF_INSTR | pl330::manager::OPERAND_INVALID |
+                pl330::manager::DMAGO_ERR | pl330::manager::EVNT_ERR |
+                pl330::manager::INSTR_FETCH_ERR | pl330::manager::DBG_INSTR,
+};
+
+enum ftr_bits {
+    FTR_MASK = pl330::channel::UNDEF_INSTR | pl330::channel::OPERAND_INVALID |
+               pl330::channel::CH_EVNT_ERR | pl330::channel::CH_PERIPH_ERR |
+               pl330::channel::CH_RDWR_ERR | pl330::channel::MFIFO_ERR |
+               pl330::channel::ST_DATA_UNAVAILABLE |
+               pl330::channel::INSTR_FETCH_ERR |
+               pl330::channel::DATA_WRITE_ERR | pl330::channel::DATA_READ_ERR |
+               pl330::channel::DBG_INSTR | pl330::channel::LOCKUP_ERR,
+};
+
+enum csr_bits {
+    DMAWFP_B_NS = bit(14),
+    DMAWFP_PERIPH = bit(15),
+    CNS = bit(21),
+};
+typedef field<0, 4, u32> CSR_CHANNEL_STATUS;
+typedef field<4, 5, u32> CSR_WAKEUP_NUMBER;
+
+enum ccr_bits {
+    SRC_INC = bit(0),
+    DST_INC = bit(14),
+};
+typedef field<1, 3, u32> CCR_SRC_BURST_SIZE;
+typedef field<4, 4, u32> CCR_SRC_BURST_LEN;
+typedef field<8, 3, u32> CCR_SRC_PROT_CTRL;
+typedef field<11, 3, u32> CCR_SRC_CACHE_CTRL;
+typedef field<15, 3, u32> CCR_DST_BURST_SIZE;
+typedef field<18, 4, u32> CCR_DST_BURST_LEN;
+typedef field<22, 3, u32> CCR_DST_PROT_CTRL;
+typedef field<25, 3, u32> CCR_DST_CACHE_CTRL;
+typedef field<28, 3, u32> CCR_ENDIAN_SWAP_SIZE;
+
+typedef field<0, 8, u32> LCO_LOOP_COUNTER_ITERATIONS;
+typedef field<0, 8, u32> LC1_LOOP_COUNTER_ITERATIONS;
+
+enum dbgstatus_bits {
+    DBGSTATUS = bit(0),
+};
+
+enum dbgcmd_bits {
+    DBGCMD = bit(0),
+};
+
+enum dbginst0_bits {
+    DEBUG_THREAD = bit(0),
+};
+typedef field<8, 3, u32> DBGINST0_CHANNEL_NUMBER;
+typedef field<16, 8, u32> DBGINST0_INSTRUCTION_BYTE0;
+typedef field<24, 8, u32> DBGINST0_INSTRUCTION_BYTE1;
+
+typedef field<0, 8, u32> DBGINST1_INSTRUCTION_BYTE2;
+typedef field<8, 8, u32> DBGINST1_INSTRUCTION_BYTE3;
+typedef field<16, 8, u32> DBGINST1_INSTRUCTION_BYTE4;
+typedef field<24, 8, u32> DBGINST1_INSTRUCTION_BYTE5;
+
+enum cr0_bits {
+    PERIPH_REQ = bit(0),
+    BOOT_EN = bit(1),
+    MGR_NS_AT_RST = bit(2),
+};
+typedef field<4, 3, u32> CR0_NUM_CHNLS;
+typedef field<12, 5, u32> CR0_NUM_PERIPH_REQ; // Number of peripheral requests
+typedef field<17, 5, u32> CR0_NUM_EVENTS;     // Number of interrupt outputs
+
+typedef field<0, 3, u32> CR1_I_CACHE_LEN;
+typedef field<4, 4, u32> CR1_NUM_I_CACHE_LINES;
+
+typedef field<0, 3, u32> CRD_DATA_WIDTH;        // data bus width of axi master
+typedef field<4, 3, u32> CRD_WR_CAP;            // TODO what does this do?
```

Comment:
Should the manual not describe what it does?

Reply:
"Write issuing capability that programs the number of outstanding write transactions"
This is what it says and I don't understand what that's supposed to mean.